lib/unleash/constraint.rb

```diff
@@ -107,7 +107,7 @@ def matches_constraint?(context)
       context_value = context.get_by_name(self.context_name)
 
       v.map!(&:upcase) if self.case_insensitive
-      context_value = context_value.upcase if self.case_insensitive
+      context_value = context_value.upcase if self.case_insensitive && context_value.is_a?(String)
```

Comment:
Talking about in place transformations.... 

I suspect that a subsequent *case sensitive* test, after a case *insensitive test*, will not return the expected result, as the underlying data has been lost due to the in place transformation...

I can try to investigate later.... But the thought just came to me now. :see_no_evil:

Reply:
> maybe we shouldn't just return false if compare using string operators, if the provided context is not a string...

We can validate what other SDKs are doing, we're aiming for consistency over correctness at the moment. I'll take some time to collect that data, because that idea also crossed my mind, and I believe the correct thing would be to return false if not a string...

> What do you think? 

I like the second alternative the most.

> I suspect that a subsequent case sensitive test, after a case insensitive test, will not return the expected result, as the underlying data has been lost due to the in place transformation...

Sorry, not a ruby expert (more like a newbie here). What's the in place transformation? Is the `is_a` operation modifying the context_value type? Or is it the `upcase`? because that was already there, maybe I didn't pay much attention but I have the feeling it's the latter...

Reply:
> > maybe we shouldn't just return false if compare using string operators, if the provided context is not a string...
> 
> We can validate what other SDKs are doing, we're aiming for consistency over correctness at the moment. I'll take some time to collect that data, because that idea also crossed my mind, and I believe the correct thing would be to return false if not a string...
> 

Sounds very wise!

> 
> Sorry, not a ruby expert (more like a newbie here). What's the in place transformation? Is the `is_a` operation modifying the context_value type? Or is it the `upcase`? because that was already there, maybe I didn't pay much attention but I have the feeling it's the latter...

Everyone should feel welcome to contribute independent of their ruby proficiency! We are here to help each other out!

`is_a?` is a safe method.  So it is the latter. And I think it was a bug from before, not a new bug.

But a bit of background, all core Ruby methods with a `!` are to be considered dangerous:

See https://docs.ruby-lang.org/en/3.1/syntax/methods_rdoc.html#label-Method+Names

>> The bang methods (`!` at the end of the method name) are called and executed just like any other method. However, by convention, a method with an exclamation point or bang is considered dangerous. In Ruby’s core library the dangerous method implies that when a method ends with a bang (`!`), it indicates that unlike its non-bang equivalent, permanently modifies its receiver. Almost always, the Ruby core library will have a non-bang counterpart (method name which does NOT end with `!`) of every bang method (method name which does end with !) that does not modify the receiver. 

doing:
```ruby
a = "Foo"
a = a.upcase
# and
a.upcase!
# are equivalent.
```

So when you try to compare them later:
```ruby
b = "Foo"

puts "a == b: #{a == b}"
```
# returns false if a `!` method was used before on top of `a`, as `a` now contains `FOO`, as the original case was lost...

Reply:
Oh thanks for the details! I think it should be doable without the in place transformation, will see if I can include that as part of this as the context should be immutable IMO

Reply:
[c2a1a09](https://github.com/Unleash/unleash-client-ruby/pull/191/commits/c2a1a09390b10bb4adc4be761728c956c78d84aa) let me know what you think. I didn't change the in place transformation because we're making a duplicate of the value some lines before: https://github.com/Unleash/unleash-client-ruby/blob/41bb8777033bbb25e58332ef7f9734a8f5ba7022/lib/unleash/constraint.rb#L106 which probably is useful to protect the context when using custom strategies (that's the only thing that comes to my mind)

Reply:
could be that the `context_value` also needs to be duplicated? (because of the `upcase` below) and not just the `v`.

a: https://github.com/Unleash/unleash-client-ruby/blob/v5.0.5/lib/unleash/constraint.rb#L106-L110

But note that this is not related to the bug fix in the scope of the PR, and I don't have a unit test confirming that this is actually a problem. So feel free to just say it is outside of the scope of this PR.